feat: implement pluggable auth interactive mode

BigTailWolf opened the pull request:
For interactive mode:
1. Always using output to read the result.
2. Make "expiration_time" optional for all mode.
3. Implement interactive mode run executable
4. Implement `revoke()` function.
4. Refactor tests

BigTailWolf committed:
feat: implementation of pluggable auth interactive mode

BigTailWolf committed:
adding test for interactive mode

lsirac commented on google/auth/pluggable.py:
```diff
@@ -132,6 +137,20 @@ def __init__(
+        ):
+            raise ValueError("Interactive timeout must be between 5 and 30 minutes.")
+        if self._interactive and not self._credential_source_executable_output_file:
+            raise ValueError("Output file must be specified in interactive mode")
```
An `output_file` must be specified in the credential configuration for interactive mode.

aeitzman reviewed (changes requested):
First pass

lsirac commented on google/auth/pluggable.py:
```diff
@@ -105,6 +106,7 @@ def __init__(
             raise ValueError(
                 "Missing credential_source. The credential_source is not a dict."
             )
+        self._interactive = os.environ.get("GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE") == "1"
```
+1 GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE is set by the auth library, not the user

lsirac commented on google/auth/pluggable.py:
```diff
@@ -183,31 +211,61 @@ def retrieve_subject_token(self, request):
-                    "Executable exited with non-zero return code {}. Error: {}".format(
-                        result.returncode, result.stdout
-                    )
+        if self._interactive:
```
Let's avoid duplicating code here

lsirac commented on google/auth/pluggable.py:
```diff
@@ -171,9 +198,10 @@ def retrieve_subject_token(self, request):
-        env[
-            "GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE"
-        ] = "0"  # Always set to 0 until interactive mode is implemented.
+        env["GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE"] = (
```
Implementation is missing GOOGLE_EXTERNAL_ACCOUNT_ID and GOOGLE_EXTERNAL_ACCOUNT_REVOKE

BigTailWolf commented on google/auth/pluggable.py:
```diff
@@ -132,6 +137,20 @@ def __init__(
         ):
             raise ValueError("Timeout must be between 5 and 120 seconds.")
 
+        if not self._credential_source_executable_interactive_timeout_millis:
```
Hi Leo, do you know where can we figure out whether it's workload?

lsirac replied:
You can check the audience

BigTailWolf replied:
Oh Thanks! I just saw the property method in the parent class.

BigTailWolf committed:
addressing comments

BigTailWolf commented on google/auth/pluggable.py:
```diff
@@ -183,31 +211,61 @@ def retrieve_subject_token(self, request):
+                            result.returncode
+                        )
+                    )
+            except Exception:
```
Fixed all kinds of this issue

BigTailWolf committed:
Merge branch 'main' into b237591436

aeitzman commented on google/auth/pluggable.py:
```diff
@@ -141,12 +157,13 @@ def retrieve_subject_token(self, request):
             raise ValueError(
                 "Executables need to be explicitly allowed (set GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES to '1') to run."
             )
+        interactive_mode = os.environ.get("GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE") == "1"
```
This still doesn't match with the design, library should be setting this environment var, not reading it.

sai-sunder-s replied:
Why not allow the user to set as well? Similar to `GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES`

aeitzman commented on google/auth/pluggable.py:
```diff
@@ -264,10 +291,11 @@ def _parse_subject_token(self, response):
             )
         if (
             "expiration_time" not in response
+            and not os.environ.get("GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE") == "1"
```
same here, don't read the environment variable.

aeitzman commented on google/auth/pluggable.py:
```diff
@@ -116,6 +117,9 @@ def __init__(
             "timeout_millis"
         )
+        self._credential_source_executable_interactive_timeout_millis = self._credential_source_executable.get(
+            "interactive_timeout_millis"
```
should there be a check that an output file has been provided if interactive timeout millis is present? Check could also happen in the retrieve_subject_token method, just think we need a clear error message if someone tries to run interactive mode with no output file.

BigTailWolf replied:
Should be in `retrieve_subject_token` since we can use both interactive mode and non-interactive mode for a config.
Then we can only check it when actually need to use the output file which is the retrieve_subject_token.

BigTailWolf commented on google/auth/pluggable.py:
```diff
@@ -171,9 +192,7 @@ def retrieve_subject_token(self, request):
-        env[
-            "GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE"
-        ] = "0"  # Always set to 0 until interactive mode is implemented.
+
```
GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE is set via `os.environ.copy()` I assume. The `os.environ` won't miss this variable

aeitzman replied:
That variable shouldn't be set already in the env, its not supposed to be a user set variable

BigTailWolf committed:
Merge branch 'main' into b237591436

BigTailWolf commented on google/auth/pluggable.py:
```diff
@@ -132,6 +137,20 @@ def __init__(
 
+        if not self._credential_source_executable_interactive_timeout_millis:
+            self._credential_source_executable_interactive_timeout_millis = (
+                5 * 60 * 1000
```
Make sense. Do you think it's good to put it in this file or somewhere else that can be referred more widely?

sai-sunder-s replied:
Lets start with this file. If we find this value repeating elsewhere, we can pull out to a common place.

BigTailWolf committed:
move interactive source of truth to kwargs in constructor and make check in the retrieve_subject_token

sai-sunder-s commented on google/auth/pluggable.py:
```diff
@@ -155,7 +174,15 @@ def retrieve_subject_token(self, request):
                     # If the cached response is expired, _parse_subject_token will raise an error which will be ignored and we will call the executable again.
                     subject_token = self._parse_subject_token(response)
+                    if (
+                        self._interactive and "expiration_time" not in response
```
Why non-interactive mode does not treat missing `expiration_time` as expired?

BigTailWolf replied:
For non-interactive mode, we already do a sanity check in the `_parse_subject_token` for `expiration_time` if an output file is specified. We treat missing expiration time as malformed and throw an exception in case of an output filed is specified.

BigTailWolf replied:
Hi @sai-sunder-s , I changed the check of "expiration_time" and unified the logic between non-interactive and interactive. This will case less confusion to users.
1. Make it as optional regardless of interactive or non-interactive
2. Treat missing expiration time as "expired" when retrieving from file

The reason we want it being optional is:
It doesn't make sense to provide an expiration time when user don't want to cache it regardless interactive or non-interactive.

aeitzman replied:
Will this make it less confusing? If we just ignore an invalid output file in non-interactive without throwing an error, the users may not realize that their output file is not being used even if they are creating it. It doesn't make sense with the current implementation for a non-interactive mode output file to not have an expiration time.

BigTailWolf replied:
Just need to document a warning "Missing expiration time will cause the file not being used for caching".
This will keep the interactive and non-interactive treat "expiration_time" in the same way.

aeitzman replied:
That would also either:
1: create a difference between the non-interactive mode behavior of the python library and the other 3 already released SDK's
2: Make the behavior of the other 3 SDK's not really make sense if we change them to match, then they wouldn't be checking for expiration_time even though they don't support interactive mode.

BigTailWolf replied:
OK. Make sense. I forget we have other libraries already made that check.

BigTailWolf committed:
implement revoke method and refactor some constants

BigTailWolf committed:
adding 2.7 check in revoke()

BigTailWolf committed:
fix lint

BigTailWolf committed:
chore: update token

aeitzman commented on google/auth/pluggable.py:
```diff
@@ -183,32 +222,87 @@ def retrieve_subject_token(self, request):
+        )
+
+        if result.returncode != 0:
+            raise exceptions.RefreshError("Auth revoke failed on executable.")
```
Include return code in the exception

BigTailWolf committed:
include error return code in exception

BigTailWolf reviewed (commented):
Required changes addressed

BigTailWolf committed:
unify the expiration_time check behavior.

BigTailWolf committed:
fix lint

BigTailWolf committed:
addressing comments

BigTailWolf committed:
chore: update token

BigTailWolf committed:
adding environment injection for revoke

BigTailWolf committed:
Revert "unify the expiration_time check behavior."

This reverts commit f361340defa2f00fe2fb33a58fa50ff908e731ff.

aeitzman commented on google/auth/pluggable.py:
```diff
@@ -183,32 +224,98 @@ def retrieve_subject_token(self, request):
+        env["GOOGLE_EXTERNAL_ACCOUNT_AUDIENCE"] = self._audience
+        env["GOOGLE_EXTERNAL_ACCOUNT_TOKEN_TYPE"] = self._subject_token_type
+        env["GOOGLE_EXTERNAL_ACCOUNT_ID"] = self.service_account_email
+        env["GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE"] = "1"
```
Its a little confusing to me that we set the interactive flag here(which makes sense sense since its only available in interactive mode) but we don't actually pipe the stdin/stdout. should make sure to clarify in the docs what we expect the executables to do when both flags are set.

BigTailWolf replied:
Added stdin/stdout since it might be needed from the executable.

BigTailWolf committed:
adding stdin/out to revoke subprocess in case some prompts may needed

BigTailWolf committed:
unifying the behavior of reading response for non-interactive and interactive

BigTailWolf committed:
adding the logic of getting external account id

BigTailWolf committed:
update token

BigTailWolf committed:
explicity inject subprocess environment variables

aeitzman commented on google/auth/pluggable.py:
```diff
@@ -212,18 +212,15 @@ def retrieve_subject_token(self, request):
-                "GOOGLE_EXTERNAL_ACCOUNT_OUTPUT_FILE"
-            ] = self._credential_source_executable_output_file
+        env["GOOGLE_EXTERNAL_ACCOUNT_IMPERSONATED_EMAIL"] = (
+            self.service_account_email or ""
```
why was this switched to filling with "" vs just not setting it at all?

BigTailWolf replied:
Redundant change by an unsophisticated idea. I'll revert.

BigTailWolf committed:
using a dummy value instead of None as a temporary solution

BigTailWolf committed:
fix environment variable injection

BigTailWolf committed:
chore: update token

lsirac commented on google/auth/pluggable.py:
```diff
@@ -183,32 +226,112 @@ def retrieve_subject_token(self, request):
+                )
+            )
+
+        # TODO: clear cache when the in memory cache feature implemented.
```
I think we should remove this comment, as we currently don't have plans to cache the response in memory.

lsirac commented on google/auth/pluggable.py:
```diff
@@ -183,32 +226,112 @@ def retrieve_subject_token(self, request):
+        subject_token = self._parse_subject_token(response)
         return subject_token
 
+    def revoke(self, request):
```
Let's refactor this to re-use the existing code that handles calling the executable

BigTailWolf replied:
In the original design doc, I saw it was supposed to be called in the form of
```
credentials.revoke()
```

By "re-use the existing code" we are calling
```
credentials.retrieve_subject_token()
```
looks like doesn't make sense literally.

lsirac replied:
Sure, but the code can be modified to be shared by the two code paths?

BigTailWolf replied:
You mean `retrieve_subject_token()` and `revoke()` share a same function of calling subprocess?

BigTailWolf replied:
I don't think that's common enough. They have different variables to populate, different stdin/stdout to assign and different post process( `_parse_subject_token` vs `_validate_response`)

BigTailWolf replied:
Adding some function to let `retrieve_subject_token()` and `revoke()` to share.

lsirac commented on google/auth/pluggable.py:
```diff
@@ -183,32 +226,112 @@ def retrieve_subject_token(self, request):
+
+    @property
+    def external_account_id(self):
+        """Get the GOOGLE_EXTERNAL_ACCOUNT_ID which needs to be polulated to executable
```
I think you should reword this here with a lot less detail and maybe explain that this should not be called. 

Something like: 

"Returns the external account identifier. When service account impersonation is used the identifier is the service account email. Without service account impersonation, this returns None, unless it is being used by the Google Cloud CLI which populates this field."

lsirac commented on google/auth/pluggable.py:
```diff
@@ -183,32 +226,112 @@ def retrieve_subject_token(self, request):
+            "GOOGLE_EXTERNAL_ACCOUNT_OUTPUT_FILE"
+        ] = self._credential_source_executable_output_file
+
+        result = subprocess.run(
```
I think you need to validate the response?

BigTailWolf replied:
Added a validation function

BigTailWolf committed:
addressing comments

BigTailWolf committed:
fix lint

BigTailWolf committed:
chore: update token

BigTailWolf committed:
refactor the common code between retrieve_subject_token and revoke

BigTailWolf committed:
chore: update token

BigTailWolf committed:
Merge branch 'main' into b237591436

lsirac commented on tests/test_pluggable.py:
```diff
@@ -316,6 +380,39 @@ def test_retrieve_subject_token_oidc_jwt(self):
             assert subject_token == self.EXECUTABLE_OIDC_TOKEN
 
+    @mock.patch.dict(os.environ, {"GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES": "1"})
+    def test_retrieve_subject_token_oidc_jwt_interactive_mode(self, tmpdir):
```
You can probably just merge this with the test above using a for loop (loop over output file responses).

BigTailWolf replied:
Refactoring to have all successful paths together will non-interactive and interactive together with all types of supported token type.

lsirac commented on tests/test_pluggable.py:
```diff
@@ -316,6 +380,39 @@ def test_retrieve_subject_token_oidc_jwt(self):
             assert subject_token == self.EXECUTABLE_OIDC_TOKEN
 
+    @mock.patch.dict(os.environ, {"GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES": "1"})
+    def test_retrieve_subject_token_oidc_jwt_interactive_mode(self, tmpdir):
```
You can probably just merge these tests into one by looping over the output file responses.

lsirac commented on tests/test_pluggable.py:
```diff
@@ -334,6 +431,38 @@ def test_retrieve_subject_token_saml(self):
 
             assert subject_token == self.EXECUTABLE_SAML_TOKEN
 
+    @mock.patch.dict(os.environ, {"GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES": "1"})
```
Might also be worth it to add a test that tests changing between interactive/non-interactive modes, but not sure what you're able to validate given the way the code is written

BigTailWolf replied:
Doing some refactoring

lsirac committed:
Merge branch 'main' into b237591436

BigTailWolf commented on docs/user-guide.rst:
```diff
@@ -450,7 +450,8 @@ Response format fields summary:
 or ``saml_response``.
-If output file is specified, ``expiration_time`` is mandatory.
+``expiration_time`` is optional. Missing ``expiration_time`` during retrieve
+from file will be treat as "expired".
```
Will do

BigTailWolf committed:
refactoring tests

BigTailWolf committed:
chore: update token

BigTailWolf committed:
Merge branch 'main' into b237591436

BigTailWolf committed:
refactor tests

clundin25 committed:
feat: Retry behavior (#1113)

* feat: Retry behavior

* Introduce `retryable` property to auth library exceptions. This can be
  used to determine if an exception should be retried.
* Introduce `should_retry` parameter to token endpoints. If set to `False`
  the auth library will not retry failed requests. If set to `True` the
  auth library will retry failed requests. The default value is `True`
  to maintain existing behavior.
* Expanded list of HTTP Status codes that will be retried.
* Modified retry behavior to use exponential backoff.
* Increased default retry attempts from 2 to 3.

BigTailWolf committed:
Revert "feat: Retry behavior (#1113)"

This reverts commit 06fff40ad9050651024e7b61cbfc5da22fa27f9b.

BigTailWolf committed:
Merge branch 'main' into b237591436

BigTailWolf committed:
chore: update token

BigTailWolf committed:
Merge branch 'main' into b237591436

BigTailWolf committed:
chore: update token

BigTailWolf committed:
Merge branch 'main' into b237591436

BigTailWolf committed:
Merge branch 'main' into b237591436

BigTailWolf committed:
chore: update token
